CenterForOpenScience/osf.io
# [feature] Fork Fixes [OSF-5652]

#### Purpose
- When forking large projects on the OSF (i.e. RPP), forks sometimes "fail" due to the request timing out after 60 seconds. However, even though the "Error: Fork Failed" growl box appears in these cases, the fork actually ends up being successful. This PR handles request timeouts so that they are not misrepresented as failures, and makes various other changes to the fork UX. 
- If a large fork times out, the user will see a "Sorry, this seems to be taking awhile" growl box that instructs them to check back for the fork later, or to contact support if the fork never appears. 
- When a fork is created successfully, the user will be given the option to go to the new forked project page or continue working where they are, much like what happens when a component is created successfully. Previously, the user was automatically redirected to the new fork. 
- Previously, users encountered several messages regarding add-on authorization of the project they were attempting to fork (the pre-fork and post-fork add-on messages were nearly the same). The before and after fork messages have been consolidated into the messages that appear before creating the fork, and the after fork messages have been removed.

#### Changes
- Forks are now created using APIv2. 
  - APIv1 routes, view functions, and related tests have been removed.
- Removes the automatic redirect to the new fork page upon success.
  - Adds a "Go to fork" or "Keep working here" modal upon success. 
- If a fork times out, the timeout error is caught and an appropriate message is displayed to the user. 
- Removes the add-on status messages that appeared on the new fork page after redirect.
  - Updated the language that appears in the "Are you sure you want to fork this project?" modal so that there is no loss of information. 

#### Side effects
- Large forks may still fail, and this PR does not attempt to fix that, but attempts to clear up the confusion surrounding large forks that timeout. 

#### QA Notes
- This PR changes the forking process to use APIv2, and removes all of the fork-related APIv1 code, and therefore fork functionality should be tested thoroughly.
  - Also, forks of projects that have various add-on configurations (fully configured, added but not configured, configured with a different user's credentials) should be tested. 
- In order to encounter a timeout error on staging, attempt to fork the largest project (the more children, logs, files on the node, the longer it will take) you can. You should receive the "Sorry this is taking awhile" message, but the fork should still be successful and should appear in the forks list soon after.
- If a large fork fails, but does not timeout, please make a separate JIRA ticket. 

#### GIFs and Screenshots
- Before Fork Modal (before)
![screen shot 2016-11-14 at 5 03 50 pm](https://cloud.githubusercontent.com/assets/7913604/20284824/dcf7cbea-aa8c-11e6-91b1-9fd1a83348c2.png)

- Before Fork Modal (after with add-on language changes)
![screen shot 2016-11-14 at 5 10 47 pm](https://cloud.githubusercontent.com/assets/7913604/20285000/a37ed9b6-aa8d-11e6-98a2-0819920b85d5.png)

- Timeout Growl (after, no before)

- Successful Fork (before -- automatic redirect to fork with messages)
![before_fork_ux](https://cloud.githubusercontent.com/assets/7913604/20284825/dd37cc0e-aa8c-11e6-8e5a-36772f52ab13.gif)

- Successful Fork (after -- "Go to..." modal)
![after](https://cloud.githubusercontent.com/assets/7913604/20285112/2d279f40-aa8e-11e6-9969-bc4e14a423f0.gif)


#### Ticket
- [OSF-5652](https://openscience.atlassian.net/browse/OSF-5652)
